### ISSUE TITLE
RFC: Frequency of major releases

### ISSUE THREAD

**Issue by sgravrock** [jasmine/jasmine#1969]
tl;dr: Should major versions of Jasmine should come out every 12-18 months instead of every 3-5 years?

Jasmine has historically had a major release every 3-5 years. This is in contrast to other JS testing frameworks like Mocha (one per year) and Jest (2-3 per year on average). We've done this partly because we think users benefit from that kind of stability, but also because we observe that major version changes impose significant costs on the community. Most users never upgrade to a newer version of their testing framework. Here are the last week's NPM download counts for the jasmine-core package:

4.x: 369,358
3.x: 1,674,328
2.x: 1,426,517

This is skewed a fair bit by Protractor, which drove increased usage of 2.8.0 for years after the 3.0 release. But the pattern is clear: A lot of people don't upgrade. I see a similar pattern with Mocha, where newer releases are somewhat more popular but old ones continue to see a lot of downloads years after they've been superseded.

Some users can't upgrade because they're dependent on other tools that require a particular major version of Jasmine. For instance, most Protractor users are stuck on Jasmine 2.8 forever, and jasmine-enzyme users probably won't be able to upgrade to Jasmine 4 until they migrate away from Enzyme. Major releases are almost always bad for those users. Other users can upgrade but don't, either because they don't wont to deal with breaking changes or because they're unaware of the new major version.

On the other hand, there are costs to infrequent major releases too. We sometimes have to say "yes, but not until long after you've moved on" to feature requests. 4.0 included features that were requested or even contributed in 2018-2019. We ended up carrying support for Internt Explorer well past the point where it became a problem for contributors and maintainers. And we haven't been able to meet the expectations of users who think that support for old Node versions should only be dropped in a major release.

I'm not sure how to weigh those concerns against each other. But it's clear that a shorter major release cycle would solve some problems, and everyone else does it. So I'm proposing to put out a major release every 12-18 months unless there are no breaking changes in the queue. We would still not backport changes to older major releases except in extraordinary cases.

Pros:
* Users don't have to wait years for features that require breaking changes.
* Old platforms won't hold back development nearly as much.
* We should be able to treat loss of support for old Node versions as a breaking change.
* Can keep our dependencies and dev dependencies more current.
* Less need for contributors to understand old platforms and JS dialects (although it's unclear if this is still a problem post-4.0).
* Major upgrades are easier when they contain only a few breaking changes.

Cons:
* Creates more frequent work for neighbor projects that interoperate with Jasmine, like Karma.
* That awkward period when a new major Jasmine version is out but it isn't supported by the surrounding ecosystem yet will happen more often.
* More users would have to take a major version upgrade to get desired new features or bugfixes.
* More users would be stuck on old, unsupported versions. (But on average, those versions would be less outdated.)
* People who use third-party packages that interoperate with Jasmine would run a greater risk of being stuck if the package is abandoned.

I'd appreciate input from the community on this.

**Comment by AaronSterlingGENEICD** [comment]
Shorter release cycle.

I would prefer to have the option to upgrade more often. Choosing not to upgrade is better than being unable to upgrade. 

Also, I don't think protractor-intertia is a strong argument. If there are enough reasons to make a major change, Protractor might upgrade, or the community might move to a more modern test runner. The advantages of upgrading need to outweigh the pain, and that will only be possible if there are valuable new features.

**Comment by sgravrock** [comment]
It sounds like there's no objection to more frequent major releases. My tentative plan is to release 5.0 next spring, to roughly coincide with Node 14's end of life.
